### PR TITLE
feat(api): protocol engine InstrumentContext implementes pick_up_tip and drop_tip

### LINKED ISSUE #7326: feat(protocol-engine): ProtocolEngine InstrumentContextInterface implements pick up tip
# Overview 

The `ProtocolEngine` accepts a `PickUpTipRequest` command. 

# Acceptance Criteria

- [ ] `pick_up_tip` method in `InstrumentContextInterface` derived class created in #7324 sends `PickUpTipRequest` command to `ProtocolEngine`

### LINKED ISSUE #7327: feat(protocol-engine): ProtocolEngine InstrumentContextInterface implements drop tip
# Overview 

The `ProtocolEngine` accepts a `DropTipRequest` command. 

# Acceptance Criteria

- [ ] `drop_tip` method in `InstrumentContextInterface` derived class created in #7324 sends `DropTipRequest` command to `ProtocolEngine`

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

Implement pick up and drop tip commands in experimental Protocol Engine `InstrumentContext`.

closes #7326 
closes #7327 

# Changelog

- Added just enough of `Well` and `Labware` to get this to work.
- Added `SyncClient.pick_up_tip` 
- Added `SyncClient.drop_tip` 

# Review requests

# Risk assessment

None

### CODE REVIEW

File: api/src/opentrons/protocol_api_experimental/instrument_context.py [hunk 1]
```diff
@@ -26,6 +27,11 @@
 class InstrumentContext:
     # noqa: D101
 
+    def __init__(self, client: ProtocolEngineClient, pipette_id: str) -> None:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/instrument_context.py]: Question for anyone: How do we name IDs from Protocol Engine to disambiguate them from other kinds of IDs? For example, here, `pipette_id` could be confused for a pipette load name, like `"p300_single_gen2"`.

Curious if other people have had similar discomfort. Or maybe it's a non-problem.

File: api/src/opentrons/protocol_api_experimental/instrument_context.py [hunk 2]
```diff
@@ -107,14 +113,39 @@ def pick_up_tip(
             presses: Optional[int] = None,
             increment: Optional[float] = None) -> InstrumentContext:
         # noqa: D102
-        raise NotImplementedError()
+        # TODO(al, 2021-04-12): What about presses and increment? They are not
+        #  supported by PE command. They are also not supported by PD protocols
+        #  either.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/instrument_context.py:118]: I'm doing something like this for `aspirate()`'s `rate`:

```suggestion
        if (presses is not None) or (increment is not None):
            raise NotImplementedError()
```

Seems less likely to get forgotten.

File: api/src/opentrons/protocol_api_experimental/instrument_context.py [hunk 3]
```diff
@@ -107,14 +113,39 @@ def pick_up_tip(
...
+
+        return self
 
     def drop_tip(
             self,
             location: Union[types.Location, Well] = None,
             home_after: bool = True) -> InstrumentContext:
         # noqa: D102
-        raise NotImplementedError()
+        # TODO(al, 2021-04-12): What about home_after?
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/instrument_context.py:139]:
```suggestion
        if not home_after:
            raise NotImplementedError()
```

Same as comment above.

File: api/src/opentrons/protocol_api_experimental/labware.py [hunk 4]
```diff
@@ -123,6 +131,15 @@ def __hash__(self) -> int:
...
     # noqa: D101
 
+    def __init__(self, labware_id: str) -> None:
+        # noqa: D107
+        self._id = labware_id
+
+    @property
+    def id(self) -> str:
+        # noqa: D102
+        return self._id
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/labware.py]: Can we somehow mark this as not belonging to the public API?

My vision is that this should be `Labware.internals.id`. But as a lightweight solution for now, we could move it to the bottom of the class, under a comment like `# Internal stuff, not part of the public Protocol API interface`?

File: api/src/opentrons/protocol_api_experimental/protocol_context.py [hunk 5]
```diff
@@ -102,7 +102,7 @@ def loaded_instruments(self) -> typing.Dict[str, InstrumentContext]:
         raise NotImplementedError()
 
     # todo(mm, 2021-04-09): Figure out a suitable type annotation for msg.
-    def pause(self, msg=None) -> None:
+    def pause(self, msg: str = None) -> None:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/protocol_context.py]: It's weird that this isn't a mypy error, isn't it?

File: api/src/opentrons/protocol_api_experimental/protocol_context.py [hunk 6]
```diff
@@ -111,7 +111,7 @@ def resume(self) -> None:
         raise NotImplementedError()
 
     # todo(mm, 2021-04-09): Figure out a suitable type annotation for msg.
-    def comment(self, msg=None) -> None:
+    def comment(self, msg: str = None) -> None:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api_experimental/protocol_context.py]: Ditto.

**Review by SyntaxColoring — CHANGES_REQUESTED** [review]

File: api/src/opentrons/protocol_api_experimental/protocol_context.py (hunk 5 above)
Comment (sanni-t) [reply api/src/opentrons/protocol_api_experimental/protocol_context.py]: mypy doesn't check things that don't have type annotations ＼| ￣ヘ￣|／

File: api/src/opentrons/protocol_api_experimental/labware.py (hunk 4 above)
Comment (amitlissack) [reply api/src/opentrons/protocol_api_experimental/labware.py]: I'll leave this for another time. There's no guarantee that this is a public interface yet.

File: api/src/opentrons/protocol_api_experimental/protocol_context.py (hunk 5 above)
Comment (amitlissack) [reply api/src/opentrons/protocol_api_experimental/protocol_context.py]: It was an error before I added the type annotation.

**Review by mcous — APPROVED** [review]
LGTM 🏇

File: api/src/opentrons/protocol_api_experimental/labware.py (hunk 4 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api_experimental/labware.py]: Sorry—no guarantee that what is a public interface yet? The whole `Labware` class?

Comment (amitlissack) [reply api/src/opentrons/protocol_api_experimental/labware.py]: Yes. The whole `Labware` class.

**Review by sanni-t — APPROVED** [review]
🙌

File: api/src/opentrons/protocol_api_experimental/protocol_context.py (hunk 5 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api_experimental/protocol_context.py]: Right, but isn't it weird that `None` counts as a `str`, as opposed to `Optional[str]`? I thought that was part of the weird mypy configuration that we were escaping by moving to a separate top-level `protocol_api_experimental` module.

Comment (amitlissack) [reply api/src/opentrons/protocol_api_experimental/protocol_context.py]: Isn't it just an implied `Optional[str]`. Does it matter?

**Review by SyntaxColoring — APPROVED** [review]
`resource_id` move looks good. Thanks!

**Pull request merged**